controller/controlpad.js

```diff
@@ -5,25 +5,58 @@
 
 const url_arg_str = window.location.search;
 const url_params = new URLSearchParams(url_arg_str);
-const subid = url_params.get('subid');
-const box_ip = window.location.href.split('/')[2].split(':')[0];
-console.log(subid);
-const ws = new WebSocket("ws://" + box_ip + ":50079");
 
-ws.onclose = () => {
-    console.log("closing");
-}
+// WebSocket class to handle communication with the controlpad server
+class GameWebSocket {
+    /**
+     * Create a new WebSocket object
+     * @param {string} ip - The IP address of the controlpad server
+     * @param {number} subid - The subid of the controlpad server
+     * @param {number} port - The port number of the controlpad server
+     */    
+    constructor(port = 50079) {
+        this.ip = window.location.href.split('/')[2].split(':')[0];
+        this.subid = url_params.get('subid');
+        this.port = port;
+        this.socket = null;
+        this.initializeWebSocket();
+    }
+
+    // initializes the WebSocket connection and binds the event handlers
+    initializeWebSocket() {
+        this.socket = new WebSocket("ws://" + this.ip + ":" + this.port);
+        this.socket.onopen = this.onopen.bind(this);
+        this.socket.onclose = this.onclose.bind(this);
+        this.socket.onerror = this.onerror.bind(this);
+        this.socket.onmessage = this.onmessage.bind(this);
+    }
+
+    // send initial message to the controlpad server after the WebSocket connection is opened
+    // Method to send initial messages once the WebSocket connection is ready
+    sendInitialMessages() {
+        const byte_array = new Uint8Array(1);
+        byte_array[0] = this.subid;
+        this.socket.send(byte_array);
+        this.socket.send("state-request");
+    }
+    
+    // event handler for the WebSocket onopen event   
+    onopen = async (_event) => {
+        console.log("opened websocket on " + this.ip + ":" + this.port);
+        this.sendInitialMessages();
+    }
+
+    // event handler for the WebSocket onclose event
+    onclose = () => {
+        console.log("closing websocket on " + this.ip + ":" + this.port);
+    }
 
-// wait for websocket to connect
-ws.onopen = (_event) => {
-    console.log("opened websocket");
-    let byte_array = new Uint8Array(1);
-    byte_array[0] = subid;
-    ws.send(byte_array);
-    ws.send("state-request");
-    ws.onmessage = async (event) => {
+    // event handler for the Websocket onmessage event
+    onmessage = async (event) => {
+        console.log("Received message:", event.data);
```

Comment:
This can get costly for some games

I did the below in one of my controlpad.js files. Let's do something like this so we can turn on and off debug messages (and so can game devs)

var _DEBUG = false;

(async () => {
  const appModule = await import('./app.js');
  if ('DEBUG' in appModule) {
    console.log('DEBUG mode:', appModule.DEBUG);
    _DEBUG = appModule.DEBUG;
  }
})();

--
ws.onclose = () => {
    if (_DEBUG) {
        console.log("websocket closed");
    }
};
--
etc. etc.